exporters/otlp/src/otlp_grpc_client.cc

```diff
@@ -46,16 +129,150 @@ static std::string GetFileContentsOrInMemoryContents(const std::string &file_pat
+  call_data->response = google::protobuf::Arena::Create<ResponseType>(call_data->arena.get());
+
+  if (call_data->request == nullptr || call_data->response == nullptr)
+  {
+    OTEL_INTERNAL_LOG_ERROR("[OTLP GRPC Client] ERROR: Export "
+                            << export_data_count << " " << export_data_name
+                            << " failed, create gRPC request/response failed");
+
+    if (call_data->result_callback)
+    {
+      call_data->result_callback(
+          opentelemetry::sdk::common::ExportResult::kFailure, std::move(call_data->arena),
+          nullptr == call_data->request ? request : *call_data->request, call_data->response);
+    }
+
+    return opentelemetry::sdk::common::ExportResult::kFailure;
+  }
+  call_data->grpc_context.swap(context);
+
+  call_data->grpc_async_callback = [](OtlpGrpcAsyncCallDataBase *base_call_data) {
+    OtlpGrpcAsyncCallData<RequestType, ResponseType> *real_call_data =
+        static_cast<OtlpGrpcAsyncCallData<RequestType, ResponseType> *>(base_call_data);
+    if (real_call_data->result_callback)
+    {
+      return real_call_data->result_callback(real_call_data->export_result,
+                                             std::move(real_call_data->arena),
+                                             *real_call_data->request, real_call_data->response);
+    }
+
+    return true;
+  };
+
+  ++async_data->start_request_counter;
+  ++async_data->running_requests;
+#  if defined(GRPC_CPP_VERSION_MAJOR) && \
+      (GRPC_CPP_VERSION_MAJOR * 1000 + GRPC_CPP_VERSION_MINOR) >= 1039
+  stub->async()
+#  else
+  stub->experimental_async()
+#  endif
+      ->Export(call_data->grpc_context.get(), call_data->request, call_data->response,
+               [call_data, async_data](::grpc::Status grpc_status) {
+                 --async_data->running_requests;
+                 ++async_data->finished_request_counter;
+
+                 call_data->grpc_status = grpc_status;
+                 if (call_data->grpc_status.ok())
+                 {
+                   call_data->export_result = opentelemetry::sdk::common::ExportResult::kSuccess;
+                 }
+
+                 if (call_data->grpc_async_callback)
+                 {
+                   call_data->grpc_async_callback(call_data.get());
+                 }
+
+                 // Maybe wake up blocking DelegateAsyncExport() call
+                 async_data->session_waker.notify_all();
+               });
+
+  // Can not cancle when start the request
+  {
+    std::unique_lock<std::mutex> lock{async_data->session_waker_lock};
+    async_data->session_waker.wait_for(lock, async_data->export_timeout, [async_data]() {
+      return async_data->running_requests.load(std::memory_order_acquire) <=
+             async_data->max_concurrent_requests;
+    });
+  }
+
+  return opentelemetry::sdk::common::ExportResult::kSuccess;
+}
+#endif
 }  // namespace
 
+OtlpGrpcClient::OtlpGrpcClient()
+#ifdef ENABLE_ASYNC_EXPORT
+    : is_shutdown_
```

Comment:
style: `is_shutdown(false)` will probably work better for clang-format.

Reply:
Good point.